cmip6_downscaling/workflows/utils.py

```diff
@@ -346,3 +350,120 @@ def write_dataset(ds: xr.Dataset, path: str, chunks_dims: Tuple = ('time',)) ->
+    connection_string : str
+        Connection string to give you write access
+    max_mem : str
+        The max memory you want to allow for a chunk. Probably want it to be around 100 MB, but that
+        is also controlled by the `calc_auspicious_chunk_sizes` calls.
+    overwrite : bool
+        Whether to overwrite the content saved at output_path if the content did not pass schema check.
+
+    Returns
+    -------
+    rechunked_ds : xr.Dataset
+        Rechunked dataset
+    """
+    # determine the chunking schema
+    if type(chunk_dims) == tuple:
+        chunk_def = calc_auspicious_chunks_dict(zarr_array, chunk_dims=chunk_dims)
+        chunks_dict = {
+            'time': None,  # write None here because you don't want to rechunk this array
+            'lon': None,
+            'lat': None,
+        }
+        for var in zarr_array.data_vars:
+            chunks_dict[var] = chunk_def
+
+    elif type(chunk_dims) == dict:
+        chunks_dict = chunk_dims
+        # ensure that the chunks_dict looks the way you want it to as {variable: {'lat': chunk_size_lat, 'lon': chunk_size_lon, 'time': chunk_size_lon}
+        # 'lon': None, 'lat': None, 'time': none}
+        for var in zarr_array.data_vars:
+            assert var in chunks_dict
+            for dim in ['lat', 'lon', 'time']:
+                chunks_dict[dim] = None
+                assert dim in chunks_dict[var]
+
+    # make the schema for what you want the rechunking routine to produce
+    # so that you can check whether what you passed in (zarr_array) already looks like that
+    # if it does, you'll skip the rechunking!
+    schema_dict = {}
+    for var in zarr_array.data_vars:
+        schema_dict[var] = DataArraySchema(chunks=chunks_dict[var])
+    target_schema = DatasetSchema(schema_dict)
+
+    # make storage patterns
+    temp_store, target_store, target_path = make_rechunker_stores(connection_string, output_path)
+
+    # check and see if the output is empty, if there is content, check that it's chunked correctly
+    if len(target_store) > 0:
+        output = xr.open_zarr(target_store)
+        try:
+            # if the content in target path is correctly chunked, return
+            target_schema.validate(output)
+            return output, target_path
+
+        except SchemaError:
+            if overwrite:
+                target_store.clear()
+            else:
+                raise NotImplementedError(
+                    'The content in the output path is incorrectly chunked, but overwrite is disabled.'
+                    'Either clear the output or enable overwrite by setting overwrite=True'
+                )
+
+    # process the input zarr array
+    delete_chunks_encoding(zarr_array)
+    try:
+        # now check if the input is already correctly chunked. If so, save to the output location and return
```

Comment:
why would we want to do this step? this seems like an unnecessary copy (unless we want to do this just for convenient/organized data storage options)

Reply:
ah, i was thinking then we don't need to rechunk when some other functions are looking for this dataset in this chunking strategy. Maybe we can add an option in the function call `cache` and only saves an extra copy if it's true?